appcues/src/main/java/com/appcues/statemachine/StateMachine.kt

```diff
@@ -90,59 +96,55 @@ internal class StateMachine(
     }
 
     private suspend fun handleActionInternal(action: Action): ResultOf<State, Error> {
-        val transition = _state.take(action)
-        val state = transition.state
-        val sideEffect = transition.sideEffect
+        return _state.takeTransition(action).run {
 
-        if (state != null) {
-            // update current state
-            _state = state
+            if (state != null) {
+                // transition to new state
+                _state = state
 
-            if (transition.emitStateChange) {
                 // emit state change to all listeners via flow
-                _stateFlow.emit(state)
+                if (emitStateChange) _stateFlow.emit(state)
```

Comment:
getState could get the last state from the shared flow ( can dig into that) and then we have only one source of truth which is the SharedFlow.

yes. I saw that we either fail and return Idling or we do null so that state is unchanged, but emiting the same state again does nothing since SharedFlow distinguishUntilChanged by nature, (surely need testing too)